connectivity/check/features.go

```diff
@@ -0,0 +1,217 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright 2022 Authors of Cilium
+
+package check
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"github.com/cilium/cilium-cli/defaults"
+	"github.com/cilium/cilium-cli/k8s"
+	"github.com/cilium/cilium/api/v1/models"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// Feature is the name of a Cilium feature (e.g. l7-proxy, cni chaining mode etc)
+type Feature string
+
+const (
+	FeatureCNIChaining        Feature = "cni-chaining"
+	FeatureMonitorAggregation Feature = "monitor-aggregation"
+	FeatureL7Proxy            Feature = "l7-proxy"
+	FeatureHostFirewall       Feature = "host-firewall"
+)
+
+// FeatureStatus describes the status of a feature. Some features are either
+// turned on or off (c.f. Enabled), while others additionally might include a
+// Mode string which provides more information about in what mode a
+// particular feature is running ((e.g. when running with CNI chaining,
+// Enabled will be true, and the Mode string will additionally contain the name
+// of the chained CNI).
+type FeatureStatus struct {
+	Enabled bool
+	Mode    string
+}
+
+func (s FeatureStatus) String() string {
+	str := "Disabled"
+	if s.Enabled {
+		str = "Enabled"
+	}
+
+	if len(s.Mode) == 0 {
+		return str
+	}
+
+	return fmt.Sprintf("%s:%s", str, s.Mode)
+}
+
+// FeatureSet contains the status
+type FeatureSet map[Feature]FeatureStatus
+
+// MatchRequirements returns true if the FeatureSet fs satisfies all the
+// requirements in reqs
+func (fs FeatureSet) MatchRequirements(reqs ...FeatureRequirement) bool {
+	for _, req := range reqs {
+		status := fs[req.feature]
+		if req.requiresEnabled && (req.enabled != status.Enabled) {
+			return false
+		}
+		if req.requiresMode && (req.mode != status.Mode) {
+			return false
+		}
+	}
+
+	// empty requirement list matches any feature set
+	return len(reqs) == 0
+}
```

Comment:
Will this ever return `true` for non-empty `reqs` where required features are enabled and in the required mode?

Reply:
No, it will not, that's a bug. Will address, thanks!

Reply:
Would be nice to add a unit test while at it.